Add CI staging job testing Ubuntu 24.04 (noble)

zenmonkeykstop opened the pull request:
## Status
Ready for review

## Description of Changes

Fixes #7312 

- adds libvirt-staging-noble molecule scenario
- updates staging CI to run against both Focal and Noble
- fixes most of the failing staging tests, rationales are in the commit messages


## Testing

- [ ] Visual review
- [ ] A staging(noble) job runs in addition to CI, and completes the `testinfra` tests.
- [ ] CI is green, esp. the two staging jobs

zenmonkeykstop committed:
Add Ubuntu Noble molecule scenario

zenmonkeykstop committed:
Update GCE CI image to version including bento/ubuntu-24.04 box

zenmonkeykstop committed:
parameterise staging CI job for ubuntu version

legoktm committed:
Ignore apparmor failures caused by ubuntu_pro_apt_news

This is an upstream Ubuntu bug that was fixed, but occurs before
packages are updated and SecureDrop is installed, so there's really
nothing for us to do until new VM images and installer ISOs are
made available

Fixes #7385.

legoktm commented:
Remaining staging failures:
* [ ] There's no longer apparmor profiles for `sbin/dhclient` and `/usr/lib/connman/scripts/dhclient-script` because they're no longer shipped in noble
* [ ] same with `/usr/sbin/tcpdump` I think?
* [ ] `test_interface_up` is failing against the source interface
* [ ] `ip6tables -S` is exiting with a zero status code despite IPv6 being fully disabled at the kernel level. Might be a nftables change or Linux 6.6. Maybe there's a better way to verify IPv6 is fully disabled?
* [ ] paxtest outputs different results

legoktm committed:
Simplify apparmor profile tests

Instead of maintaining the same version-specific list of profiles six
times, just check against a single list of profiles we care about,
namely tor and apache2. The rest are not something under our control and
vary based on the Ubuntu version.

test_apparmor_ensure_not_disabled was mostly broken, because it was
looking for files like
`/etc/apparmor.d/disabled/usr.sbin./usr/sbin/tor`, which of course would
never exist. Instead just check that there are no disable files
installed.

legoktm committed:
Improve test_grsecurity_paxtest; temporarily disable on noble

* Remove the unnecessary diffing and templating, just keep what we
  want to assert against as a plain string.
* Avoid shelling out to grep, we can do the filtering in Python.
* Add the expected noble output, but disable the test because it
  currently isn't passing with unclear output.

legoktm commented:
I've just been pushing stuff here for now as I work through the failures, later we can figure out if we want them to get individual PRs or not.

legoktm committed:
Fix ip6tables tests on noble

* ip6tables no longer errors if IPv6 is disabled, so limit that test to
  focal only.
* Add a new test that verifies IPv6 is disabled at the kernel level by
  looking at the relevant /proc file.

legoktm commented:
An interesting quirk of ip6tables on noble; despite IPv6 being disabled the command still works, so on boot you get a wide-open ruleset of:
```
sdadmin@mon$ sudo ip6tables -S
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
```

This is because netfilter-persistent is smart and checks if IPv6 is disabled, and if so, doesn't load rules. So our "drop everything" in `install_files/ansible-base/roles/restrict-direct-access/files/iptables_rules_v6` is never even loaded. We should still keep it around as a defense-in-depth just in case IPv6 is ever actually turned on, it'll get loaded and drop everything.

legoktm committed:
Ensure paxtest gets installed

Don't let unattended-upgrades interfere with the installation by holding the
dpkg lock.

legoktm commented:
I fixed the paxtest failure on focal because of unattended-upgrades preventing it from being installed.

The output on noble is uh, off. On app-staging, it's reliably:
```
...
Return to function (strcpy)              : paxtest: return address contains a NULL byte.
Return to function (memcpy)              : Killed
Return to function (strcpy, PIE)         : paxtest: return address contains a NULL byte.
Return to function (memcpy, PIE)         :
```
(the last line doesn't have a result)

And then on mon-staging:
```
...
Return to function (strcpy)              : paxtest: return address contains a NULL byte.
Return to function (memcpy)              : Return to function (strcpy, PIE)         : paxtest: return address contains a NULL byte.
Return to function (memcpy, PIE)         :
```
(interspersed output, missing two results)

legoktm commented:
@zenmonkeykstop I'm curious if you have any ideas on why the paxtest output is different across hosts and how to design a test to address it?

zenmonkeykstop commented:
> @zenmonkeykstop I'm curious if you have any ideas on why the paxtest output is different across hosts and how to design a test to address it?

Not off the top of my head -  assuming same kernel and paxctld setup the results should be the same. I'll look into it once we're done with 2025JanPlanfest.

legoktm commented:
Ack; I've set the test to be skipped on noble so we can get a green check and land this and fix it incrementally instead of holding the entire staging job back - I've flipped it out of draft mode.

The individual commit messages have rationales for all the further changes.

zenmonkeykstop reviewed (commented):
@legoktm your changes LGTM - I can't merge coz original author, will flag in channel

legoktm reviewed (approved):
Per Kev's review
